core/src/main/java/org/elasticsearch/cockroach/TransportCockroachAction.java

```diff
@@ -0,0 +1,115 @@
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.function.Supplier;
+
+/**
+ * An action that can survive restart of requesting or executing node.
+ * These actions are using cluster state rather than only transport service to send requests and responses.
+ */
+public abstract class TransportCockroachAction<Request extends CockroachRequest<Request>, Response extends CockroachResponse>
+    extends HandledTransportAction<Request, Response> {
+
+    private final CockroachService cockroachService;
+
+    private final String executor;
+
+    protected TransportCockroachAction(Settings settings, String actionName, boolean canTripCircuitBreaker, ThreadPool threadPool,
+                                       TransportService transportService, CockroachService cockroachService, ActionFilters actionFilters,
+                                       IndexNameExpressionResolver indexNameExpressionResolver,
+                                       Supplier<Request> requestSupplier, Supplier<Response> responseSupplier, String executor) {
+        super(settings, actionName, canTripCircuitBreaker, threadPool, transportService, actionFilters, indexNameExpressionResolver,
+            requestSupplier);
+        this.cockroachService = cockroachService;
+        this.executor = executor;
+        cockroachService.registerRequestHandler(actionName, this, requestSupplier, responseSupplier, executor);
+    }
+
+    /**
+     * Returns the node id where the request has to be executed
+     */
+    public abstract DiscoveryNode executorNode(Request request, ClusterState clusterState);
+
+    /**
+     * Returns the node id where the response should be processed
+     */
+    public abstract DiscoveryNode responseNode(Request request, ClusterState clusterState);
+
+    /**
+     * Checks the current cluster state for compatibility with the request
+     * <p>
+     * Throws an exception if the supplied request cannot be executed on the cluster in the current state.
+     */
+    public void validate(Request request, ClusterState clusterState) {
+
+    }
+
+    @Override
+    protected final void doExecute(Request request, ActionListener<Response> listener) {
+        logger.warn("attempt to execute a cockroach action without a task");
+        throw new UnsupportedOperationException("task parameter is required for this operation");
+    }
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        cockroachService.sendRequest((CockroachTask) task, actionName, request, listener);
+    }
+
+    /**
+     * This operation will be executed on the caller node.
+     */
+    public abstract void executorNodeOperation(Task task, Request request, ActionListener<Response> listener);
```

Comment:
I think the comment here should be that the operation will be executed on the executor node?